_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,359 @@
+
+```scala
+import shapeless.=:!=
+def foo[Ole](a: Ole)(implicit ev: =:!=[Ole, Int]): A = a
+
+foo[String]("E.Near rocks.")   /* Compiles */
+foo[Int](5)                    /* Does not compile */
+foo[Boolean](true)             /* Compiles */
+```
+
+Briefly, if <code>Ole</code> and <code>Int</code> are different, the compiler will be able to (always) find the required implicit and succeed the compilation. If they are the same, something else happens and a compilation error is raised. How exactly using the syntax <code>(implicit ev: =:!=[Ole, Int])</code> achieves that behavior is beyond this post; it should be accepted as the truth.
+
+This can however be used to tackle our problem:
+
+```scala
+import shapeless.=:!=
+case class Reservation[SeatType <: Seat](
+  userTicket: Ticket[SeatType],
+  guestTicket: Ticket[SeatType]
+)(implicit ev: SeatType =:!= Seat)
+```
+
+The implicit evidence **ev** will demand the compiler to find proof that <code>SeatType</code>, whatever type it might be, must not be type <code>Seat</code>. If it cannot, it throws a compilation error.  With figure 1 as guidance, because abstract type <code>SeatType</code> has as upper bound the concrete type <code>Seat</code>, and since given the generalized type constraint, <code>SeatType</code> must not be <code>Seat</code>, then it must surely be one of <code>Standing</code>, <code>VipTribune</code> or <code>NormalSeat</code>. Problem solved. So, in contrast with the other 2 solutions so far seen, an attempt to instantiate a reservation with unsound parameters is caught during compilation:
+
+```scala
+val standing = Standing
+val normalSeated = NormalSeat(4, 5)
+val tribuneSeat = VipTribune("east")
+
+//  Compiles
+val bar = Reservation(
+  Ticket("id1", 123, normalSeated),
+  Ticket("id2", 456, normalSeated)
+)
+
+//  Will never compile
+val foo = Reservation(
+  Ticket("id1", 123,  standing),
+  Ticket("id2", 456, tribuneSeat)
+)
+
+//  Will never compile
+val baz = Reservation(
+  Ticket("id1", 123, normalSeated),
+  Ticket("id2", 456, tribuneSeat)
+)
+```
+
+When it does not compile we get the error:  
+ 
+```
+ambiguous implicit values:
+[error]  both method neqAmbig1 in package shapeless of type [A]=> A =:!= A
+[error]  and method neqAmbig2 in package shapeless of type [A]=> A =:!= A
+[error]  match expected type com.github.cmhteixeira.invariants.Seat =:!= com.github.cmhteixeira.invariants.Seat
+[error] val foo = Reservation(
+[error]                           ^
+```
+
+The error message is not very helpful, but one must live with it <a id="SO-answer-error-message-union-types-ref" href="#SO-answer-error-message-union-types"> ¹ </a>
+
+This approach, similarly to the 2 other approaches, suffers from an inconvenient. The type <code>SeatType</code> on a given reservation is lost at run time due to type erasure. We cannot pattern match on a given reservation as on the following example:  
```

Comment:
> suffers from an inconvenient.

I think the adjective `inconvenient` should be replaced by the associated noun `inconvenience`.

Reply:
You are right. Changed.